File: pyJamVentSim/testing/test00.py
#!/usr/bin/env python3
#
# ultra simple test of basic gas elements.
#
# source -- pipe -- valve -- reservoire...

import sys;
import time;
import pprint as pp
import matplotlib.pyplot as plt


from GasSim.SimNode import SimNode,NodeType
from GasSim.GasSource import GasSource
from GasSim.Valve import Valve
from GasSim.Reservoir import Reservoir
from GasSim import Const


#
# gasSrc <-> valveRin <-> resovior
#  3bar        .1R         1 liter
# pressures are in absolute pressure
# temperature is 21c

# time constant check
# the end result of this after 1 second should be
#    within acceptable rounding errors...
# the same regardless of of dt
#
#

class SimpleGasModel(SimNode):
    def __init__(self,srcPres,resPres):
        SimNode.__init__(self,"simpleGasModel", NodeType.MODEL)
        # create all the nodes...
        #
        pipeResist=0.1;
        gasSrc = GasSource("gasSrc",srcPres)  #  bar air source.
        self.addChildNode(gasSrc);
        valveRin = Valve("valveRin", 1,  # pressure
                         pipeResist)  # resistance
        self.addChildNode(valveRin);
        reservoir = Reservoir("reservoir", resPres,  # pressure
                             2)  # 2 liter volume
        self.addChildNode(reservoir);

        self._gasSrc    = gasSrc;
        self._valveRin  = valveRin
        self._reservoir = reservoir

        # gasSrc <-> valveRin <-> resovior
        # then connect them up..
        gasSrc.connect(valveRin);
        valveRin.connect(gasSrc);

        valveRin.connect(reservoir);
        reservoir.connect(valveRin);

    def getGasSrc(self):
        return(self._gasSrc)
    def getValveRin(self):
        return(self._valveRin);
    def getReservoir(self):
        return(self._reservoir);

# run the model and return
# an array of resovire pressures during the run
#    for plotting and checking...
#
def doModel(dt,timeLimit, stride):
    model = SimpleGasModel(3.0,     # srcPres
                           1.0      # resPres
                           );
    gasSrc = model.getGasSrc();
    reservoir = model.getReservoir();
    valveRin = model.getValveRin();
    valveRin.setOpen(True);

    T=[];       # arrays to plot
    Pres = [];  # reservoir pressures
    nsteps=0;
    timeNow=0.0

    T.append(timeNow);
    Pres.append(reservoir.out.pressure);
    while timeNow < timeLimit:
        nsteps += 1;
        # strid the timesteps, so our number of
        #    samples are the same, so we can plot them
        #    together...
        if ((nsteps % stride) == 0):
            model.step(dt*stride);
            model.next();
        timeNow += dt;
        T.append(timeNow);
        Pres.append(reservoir.out.pressure);
    return(T,Pres)


def plotPres(X, Y, color, xMax, yMax):
    plt.plot(X, Y, color=color, linewidth=2.5, linestyle="-")
    plt.xlim(0, xMax)
    #plt.ylim(0,max(reservoir.out.pressure,gasSrc.out.pressure));
    plt.ylim(0, yMax);
    plt.draw();

if __name__ == "__main__":
    # keep this very simple
    timeLimit=4
    plt.axis([0,timeLimit,0,4])
    print("stride=10")
    (T,PresB) = doModel(0.01,          # dt
                       timeLimit,     # timelimit
                       10)             # stride
    print("stride=1")
    (T,PresA) = doModel(0.01,          # dt
                       timeLimit,     # timelimit
                       1)             # stride
    plotPres(T, PresA, "blue", timeLimit, 4)
    plotPres(T, PresB, "red", timeLimit, 4)
    plt.pause(0.001)        # force it to diplay

    # for a real test, don't do this...
    if (round(PresA[-1],6) != round(PresB[-1],6)):
        print("PresA[-1]({}) != PresB[-1]({})".format(round(PresA[-1],2),round(PresB[-1],2)))
    input("Press [enter] to continue.")

File: jamventProto/jamventsimlib/testing/test00.cpp
//
//  main.cpp
//  xcode-test01
//
//  Created by ralph bellofatto on 4/12/20.
//  Copyright © 2020 ralph bellofatto. All rights reserved.
//

#include <iostream>
#include <memory>
#include <vector>
#include <string>
#include <iomanip>
#include <iostream>
#include <strstream>
#include <istream>
#include <cmath>
#include <ostream>
#include <sstream>


#include "StrTokenize.h"
#include "fmtstr.h"
#include "GassimModel.h"

using namespace std;

typedef vector<string> ErrStrs;

// #
// # gasSrc <-> valveRin <-> resovior
// #  3bar        .1R         1 liter
// # pressures are in absolute pressure
// # temperature is 21c

// # time constant check
// # the end result of this after 1 second should be
// #    within acceptable rounding errors...
// # the same regardless of of dt
// #
// #



class SimpleGasModel: public GassimModel 
{
public:
	SimpleGasModel() {
		double pipeResist=0.1;
		double resPress = 1;
		// create the nodes
		_gasSrc = std ::make_shared<GassimSource>("gasSrc", 2.0, .21);
		_valveRin = std ::make_shared<GassimPipe>("valveRin", pipeResist, true);
		_reservoir = std::make_shared<GassimReservoir>("reservoir", resPress, 2);

		// add them to the mode.
		addNode(_gasSrc);
		addNode(_valveRin);
		addNode(_reservoir);

		// connect them up...
		connect(_gasSrc,_valveRin);
		connect(_valveRin,_reservoir);

	}
	// public container...
	GassimNode::NodePtr_t _gasSrc;
	GassimNode::NodePtr_t _valveRin;
	GassimNode::NodePtr_t _reservoir;

protected:
private:


};

typedef std::pair<double, double> TimeSeriesPair_t;
typedef std::vector<TimeSeriesPair_t> VectorSeries_t;
typedef std::shared_ptr<VectorSeries_t> VectorSeriesPtr_t;

VectorSeriesPtr_t doModel(double dt, 
	double timeLimit, 
	uint64_t stride)
{
	VectorSeriesPtr_t vp(make_shared<VectorSeries_t>());
	SimpleGasModel  model;
    uint64_t nsteps=0;
	double timeNow = 0.0;

    while (timeNow < timeLimit) {
        nsteps++;
	    // # strid the timesteps, so our number of
	    // #    samples are the same, so we can plot them
	    // #    together...
	    if((nsteps % stride) == 0) 	    {
	        model.step(dt*stride);
            model.next();
		}
	    vp->push_back(TimeSeriesPair_t(timeNow, model._reservoir->pressure()));
		timeNow += dt;

    }
	

	return(vp);
}
int  main(int argc, const char * argv []) {
	ErrStrs errs;
	bool passed = true;
	cout << "Testing test00 ..." << endl;
	
	double timelimit  = 4;
	VectorSeriesPtr_t vp = doModel(0.01, timelimit, 1);

#if 1
	// just blindly print out the vector...
	for (auto it = vp->begin(); it != vp->end(); it++) 
		cout << it->first <<  ", " << it->second << endl;
#endif
	if (errs.size() > 0) {
		for (auto iter = errs.begin(); iter != errs.end(); iter++) 
			cout << *iter << endl;
		cout << "TEST Failed" << endl;
		passed = false;
	}
	
	
	cout << "Test : " << (passed ? "PASSED" : "FAILED") << endl;

	return (passed ? 0 : 1);
}

